util/src/main/java/io/camunda/zeebe/util/FileUtil.java

```diff
@@ -20,16 +22,48 @@
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.nio.file.SimpleFileVisitor;
+import java.nio.file.StandardOpenOption;
 import java.nio.file.attribute.BasicFileAttributes;
 import java.util.Objects;
 import org.slf4j.Logger;
 
 public final class FileUtil {
-
-  public static final Logger LOG = Loggers.FILE_LOGGER;
+  private static final Logger LOG = Loggers.FILE_LOGGER;
 
   private FileUtil() {}
 
+  /**
+   * Ensures updates to the given path are flushed to disk, with the same guarantees as {@link
+   * FileChannel#force(boolean)}. This can be used for both directory and files. Note that if you
+   * already have a file channel open, then use the {@link FileChannel#force(boolean)} method
+   * directly. This method is mostly useful when flushing directories after file creation and/or
+   * file rename operations.
+   *
+   * @param path the path to synchronize
+   * @throws IOException can be thrown on opening and on flushing the file
+   */
+  public static void flush(final Path path) throws IOException {
+    try (final var channel = FileChannel.open(path, StandardOpenOption.READ)) {
+      channel.force(true);
+    }
+  }
+
+  /**
+   * Moves the given {@code source} file to the {@code target} location, flushing the target's
+   * parent directory afterwards to guarantee that the file will be visible and avoid the classic
+   * 0-length problem.
+   *
+   * @param source the file or directory to move
+   * @param target the new path the file or directory should have after
+   * @param options copy options, e.g. {@link java.nio.file.StandardCopyOption}
+   * @throws IOException on either move or flush error
+   */
+  public static void moveDurably(final Path source, final Path target, final CopyOption... options)
+      throws IOException {
+    Files.move(source, target, options);
+    flush(target.getParent());
```

Comment:
It could be necessary if the move operation ends up being a copy operation, yes - if it's a normal move, then it's only necessary to flush the parent directory since all we did was update the source's parent/target's parent directory entries (but nothing else).

I did think a bit about the copy case, but I don't think it's a common use case? It would mean you're crossing file system boundaries, which isn't really possible right now with our configuration. Of course, with certain file systems move _is_ a copy, but I hope these would also handle this case? Possibly not :sweat_smile: As it is, we keep feature parity with what we had - it was just extracted. I'm happy to discuss the copy case if you think it's more prevalent than I thought :+1: